### PR TITLE
[SYCL] Add tests for atomic operations on malloc_shared-allocated memory

### PR DESCRIPTION
**Pull request by pasaulais**
This duplicates existing SYCL atomic tests that use buffers with identical tests that use USM and memory allocated using `sycl::malloc_shared`. The motivation behind exercising atomics on `malloc_shared`-allocated memory is to catch issues such as https://github.com/ROCm/ROCm/issues/848 where native atomic instructions may fail when crossing the PCIe bus.

### CODE REVIEW

**Comment by pasaulais** [comment]
This PR so far only contains changes to the 'add' atomic tests to get initial feedback. I will add other tests later so that review feedback won't potentially require all tests to be changed.

**Comment by maarquitos14** [comment]
> This PR so far only contains changes to the 'add' atomic tests to get initial feedback. I will add other tests later so that review feedback won't potentially require all tests to be changed.

The changes so far look good to me. Only one nit: comments are supposed to be written in proper English, i.e. with proper punctuation. I noticed that you simply copy-pasted the comments from existing tests into the new tests, but can you add the `.` at the end of the sentence for each of them? Just for the newly added ones.

**Comment by pasaulais** [comment]
> > This PR so far only contains changes to the 'add' atomic tests to get initial feedback. I will add other tests later so that review feedback won't potentially require all tests to be changed.
> 
> The changes so far look good to me. Only one nit: comments are supposed to be written in proper English, i.e. with proper punctuation. I noticed that you simply copy-pasted the comments from existing tests into the new tests, but can you add the `.` at the end of the sentence for each of them? Just for the newly added ones.

No problem. I've clang-formatted my changes and added `.` to the comments in the new code. I can do it for the existing code as well  as I go. Is there any reason not to, like keeping the diff simple?

**Comment by maarquitos14** [comment]
> No problem. I've clang-formatted my changes and added `.` to the comments in the new code. I can do it for the existing code as well as I go. Is there any reason not to, like keeping the diff simple?

Yes, that's exactly it, we want to keep the diff simple.

**Comment by pasaulais** [comment]
@maarquitos14 I have completed the tests for all atomic operations present in the `test-e2e/AtomicRef` folder, using the same approach as the 'add' test you've already reviewed. Could you take another look at it?

File: sycl/test-e2e/AtomicRef/exchange.h [hunk 1]
```diff
@@ -99,6 +99,44 @@ void exchange_global_test(queue q, size_t N) {
...
+              scope, space > (exc[0]);
+         output[gid] = atm.exchange(T(gid), order);
+       });
+     }).wait_and_throw();
+  }
+
+  // Only one work-item should have received the initial sentinel value.
+  assert(std::count(output_begin, output_end, initial) == 1);
+
+  // All other values should be unique; each work-item replaces the value it.
```
[the remark this answers is not in the document]
Comment (pasaulais) [reply sycl/test-e2e/AtomicRef/exchange.h]: Good idea, done.

File: sycl/test-e2e/AtomicRef/min.h [hunk 2]
```diff
@@ -106,6 +106,48 @@ void min_global_test(queue q, size_t N) {
...
+  }
+
+  // Final value should be equal to 0.
+  assert(val[0] == 0);
+
+  // Only one work-item should have received the initial value.
+  assert(std::count(output_begin, output_end, initial) == 1);
+
+  // fetch_min returns original value.
+  // Intermediate values should all be <= initial value.
```
[the remark this answers is not in the document]
Comment (pasaulais) [reply sycl/test-e2e/AtomicRef/min.h:142]: Good idea, done.

**Comment by pasaulais** [comment]
Thanks for your comments @maarquitos14, I think I've addressed all of them.

**Comment by pasaulais** [comment]
All atomic tests now seem to be failing on CI (AMD HIP only as far as I can see) with:

```
# | terminate called after throwing an instance of 'sycl::_V1::exception'
# |   what():  Device does not support shared USM allocations!
```

**Pull request merged**